src/Storages/ExternalStream/Kafka/KafkaSink.cpp

```diff
@@ -78,8 +186,41 @@ KafkaSink::KafkaSink(const Kafka * kafka, const Block & header, ContextPtr conte
         }
     }
 
-    rd_kafka_conf_set_dr_msg_cb(conf, [](rd_kafka_t * /* producer */, const rd_kafka_message_t * msg, void * /* opaque */) {
-        static_cast<WriteBufferFromKafka *>(msg->_private)->onMessageDelivery(msg);
+    auto * topic_conf = rd_kafka_conf_get_default_topic_conf(conf);
+    if (!topic_conf)
+    {
+        topic_conf = rd_kafka_topic_conf_new();
+        /// conf will take the ownership of topic_conf
+        rd_kafka_conf_set_default_topic_conf(conf, topic_conf);
+    }
+
+    rd_kafka_conf_set_opaque(conf, this);
+    rd_kafka_topic_conf_set_opaque(topic_conf, this);
+
+    /// With partitioner callback, we can get the up-to-date partition count w/o additional effort.
+    rd_kafka_topic_conf_set_partitioner_cb(topic_conf, [](const rd_kafka_topic_t * /*rkt*/,
+						const void * /*keydata*/,
+						size_t /*keylen*/,
+						int32_t partition_count,
+						void * rkt_opaque,
+						void * msg_opaque) -> int32_t
+    {
+        /// update partition count
+        auto * sink = static_cast<KafkaSink *>(rkt_opaque);
+        sink->partition_cnt = partition_count;
+
+        auto * partition_id_ptr = static_cast<Int32 *>(msg_opaque);
+        auto parition_id =  *partition_id_ptr;
+        delete partition_id_ptr;
+        /// This should not really happen because Kafka does not support reducing partitions.
```

Comment:
If partitioned gets changed in the middle, this partition assignment make not make sense anymore, but in general, the sharding strategy might not make sense any more.